account_invoice_inter_company/models/account_invoice.py

```diff
@@ -0,0 +1,288 @@
+# -*- coding: utf-8 -*-
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError, AccessError
+from odoo.tools import float_compare
+
+
+class AccountInvoice(models.Model):
+
+    _inherit = 'account.invoice'
+
+    auto_generated = fields.Boolean(string='Auto Generated Document',
+                                    copy=False, default=False)
+    auto_invoice_id = fields.Many2one('account.invoice',
+                                      string='Source Invoice',
+                                      readonly=True, copy=False,
+                                      _prefetch=False)
+
+    @api.multi
+    def action_invoice_open(self):
+        """ Validated invoice generate cross invoice base on company rules """
+        res = super(AccountInvoice, self).action_invoice_open()
+        for src_invoice in self:
+            # do not consider invoices that have already been auto-generated,
+            # nor the invoices that were already validated in the past
+            dest_company = self.env['res.company']._find_company_from_partner(
+                src_invoice.partner_id.id)
+            if dest_company and not src_invoice.auto_generated:
+                if src_invoice.type == 'out_invoice':
```

Comment:
You can rewrite this sequence of if..elif.. in another way, for example:

- create two data structures, one that maps the four possible values source invoice types into another invoice type, the other that maps them into a journal type, something like:

```
MAP_INVOICE_TYPE = {
    'out_invoice': 'in_invoice',
    'in_invoice': 'out_invoice',
    'out_refund': 'in_refund',
    'in_refund': 'out_refund',
}
MAP_JOURNAL_TYPE = {
    'out_invoice': 'purchase',
    'in_invoice': 'sale',
    'out_refund': 'purchase_refund',
    'in_refund': 'sale_refund',
}
```

- set the values you need by using the above data structures, something like:
```
    dest_inv_type = MAP_INVOICE_TYPE.get(str_type)
    dest_journal_type = MAP_INVOICE_TYPE.get(str_type)
```


This way you will increase the coverage without adding extra tests.